### PR TITLE
Fix for module.run

### PR DESCRIPTION
**Pull request by cachedout**
Thanks to @isbm for help.

cc: @brejoc , @Ch3LL

### CODE REVIEW

File: salt/states/module.py [hunk 1]
```diff
@@ -319,13 +319,10 @@ def _call_function(name, returner=None, **kwargs):
     arg_type, na_type, kw_type = [], {}, False
```
Comment (DmitryKuzmenko) [inline salt/states/module.py]: `kw_type` is seems unneeded anymore.

File: salt/states/module.py [hunk 2]
```diff
@@ -319,13 +319,10 @@ def _call_function(name, returner=None, **kwargs):
...
         if not isinstance(funcset, dict):
-            kw_type = True
-        if kw_type:
-            if isinstance(funcset, dict):
-                arg_type += funcset.values()
-                na_type.update(funcset)
-            else:
-                arg_type.append(funcset)
+            arg_type.append(funcset)
+        elif isinstance(funcset, dict):
```
Comment (DmitryKuzmenko) [inline salt/states/module.py]: It looks `else` is enough here.

**Review by DmitryKuzmenko — CHANGES_REQUESTED** [review]

File: salt/states/module.py (hunk 1 above)
Comment (terminalmage) [reply salt/states/module.py]: Yeah, when looking at the [whole file](https://github.com/cachedout/salt/blob/cd6f0dd08c9afe633e2781fdc169cc73802b87f7/salt/states/module.py) it seems like he only code that references `kw_type` is gone.

File: salt/states/module.py (hunk 2 above)
Comment (terminalmage) [reply salt/states/module.py]: Yeah. Again, when you factor out the removed code, the preceding `if` is already doing `not isinstance(funcset, dict)`. This means that an `else` implicitly assumes that `funcset` _is_ a dict, making a second type check redundant.

**Review by terminalmage — CHANGES_REQUESTED** [review]
I agree with @DmitryKuzmenko, I've added some clarification/explanation.

File: salt/states/module.py [hunk 3]
```diff
@@ -319,13 +319,10 @@ def _call_function(name, returner=None, **kwargs):
...
+            arg_type += funcset.values()
+            na_type.update(funcset)
         else:
```
Comment (terminalmage) [inline salt/states/module.py]: Also, this `else` should be removed. The `func_kw.update(funcset)` was only happening when `kw_type` was `False`, which only happens when `funcset` is a dict. Due to the initial `isinstance` performed above the differing code, we already will know that `funcset` is a dict if we simply change the 2nd `isinstance` to an `else`.

**Review by terminalmage — CHANGES_REQUESTED** [review]

**Comment by cachedout** [comment]
@isbm I made some additional changes here. Do you have time to take a look?

**Review by terminalmage — APPROVED** [review]

**Review by DmitryKuzmenko — APPROVED** [review]

**Comment by rallytime** [comment]
@cachedout This change has caused a unit test to fail.

**Comment by terminalmage** [comment]
OK, this is a legit failure. This change is causing the argument to be added both to `arg_type` and `func_kw`, which are passed to the function as postional and keyword arguments, respectively. This results in the same variable name being passed twice.

I've tried to figure out just what on earth this code does, but unfortunately it doesn't really make any sense and doesn't have _any_ comments to explain what it does. I'm referring this back to @isbm since he wrote it.

I'll be submitting a PR to @cachedout's fork soon to improve the unit tests, as they don't really offer anything in the way of troubleshooting information when they fail.

**Comment by isbm** [comment]
@cachedout finally back to my desk, sorry for the delay. I tried to figure out again how that code landed like this, but seems like some commit in between was missed, as in our branch it is half-broken, and here is completely broken (and so we differ!). Nevertheless, thanks for the fix, as long as it is there, we will sync it all up ASAP.

Thanks for working on this during our conference!

**Comment by cachedout** [comment]
Hi @isbm. Glad you made it back. :] I have actually replaced this PR with a new version which is https://github.com/saltstack/salt/pull/43847

Do you have time to review those changes? I am going to close this PR in favor of that one.

**Pull request closed without merging**